Static event handlers inference

itaipelles opened the pull request.

ohana54 commented:
I didn't have time to fully review, but 2 things:
1. What happens when an element is deleted but the event handler remains. 
2. The whole code is very hard to read and understand, I think we need better abstractions on top of tern. 
What do you think?

itaipelles commented:
1. the events data we get from santa-editor is represented as a list of components, each with a .events property. so we can not programmatically receive events with no matching component. Event handlers that are abscent from the symbols will simply exit the function here https://github.com/wix/wix-code-ide/blob/72ebf473e5557ecd075803861bb5eeb6da83497a/src/wix-code/tern/plugins/wixStaticEventHandlers.js#L92 and won't get any type inference.

2. I did a pretty big refactor and i think it's much more readable now. Please give it another try and let's discuss after that? I think wrapping tern with an abstraction layer that makes sense is a very good idea, but i'm not sure we're ready for it yet.

itaipelles commented:
@ob1 
I implemented the changes we discussed and did some refactoring.
Also, with Sam pushing the documentation of the new $w parameter to all events, the previous solution did not exactly work.. So now i'm putting the types data of the parameters on the scope object, rather than on the function type object (why is there no single source of truth..? :( )
I also added tests for $w in static event handlers.
Can you have a look?

ohana54 commented on src/wix-code/tern/plugins/wixStaticEventHandlers.js:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    function putArgumentTypesOnMatchingPropertiesOfScope(actualHandlerFunction, ast, eventHandlerType) {
+        var actualPropNames = actualHandlerFunction.types[0].argNames;
```
`functionArgumentNamesInCode`

ohana54 commented on src/wix-code/tern/plugins/wixStaticEventHandlers.js:
```diff
@@ -0,0 +1,112 @@
+
+    function putArgumentTypesOnMatchingPropertiesOfScope(actualHandlerFunction, ast, eventHandlerType) {
+        var actualPropNames = actualHandlerFunction.types[0].argNames;
+        var functionScope = infer.scopeAt(ast, actualHandlerFunction.originNode.end);
```
`functionBodyScope`?

ohana54 commented on src/wix-code/tern/plugins/wixStaticEventHandlers.js:
```diff
@@ -0,0 +1,112 @@
+    function putArgumentTypesOnMatchingPropertiesOfScope(actualHandlerFunction, ast, eventHandlerType) {
+        var actualPropNames = actualHandlerFunction.types[0].argNames;
+        var functionScope = infer.scopeAt(ast, actualHandlerFunction.originNode.end);
+        Object.keys(functionScope.props).forEach(function(functionScopePropName) {
```
refactor to run on the actual prop names

ohana54 commented on src/wix-code/tern/plugins/wixStaticEventHandlers.js:
```diff
@@ -0,0 +1,112 @@
+
+    var componentEventTypeToPropName = null;
+
+    function postLoadDef() {
```
Remove and do it when the module loads

ohana54 commented on src/wix-code/tern/plugins/wixStaticEventHandlers.js:
```diff
@@ -0,0 +1,112 @@
+        var componentTypes = Object.keys($wNameSpace);
+        componentTypes.forEach(function(componentType) {
+            var prototype = $wNameSpace[componentType].prototype;
+            if (!prototype) {
```
why? write a comment with an example

ohana54 commented on test/browser/autocomplete/event-callbacks.spec.js:
```diff
@@ -18,29 +18,85 @@ describe('event callbacks', () => {
 
-    it('should display proposals for MouseEvent', async function() {
-        const editorLoaderDriver = editorLoaderDriverCreator(this.browser);
+    describe('dynamic', function() {
```
handle all possible cases
